shared/middle/sicp.md

```diff
@@ -0,0 +1,214 @@
+
+    * What is the closure property? Does it mean the same as in Lisp/Javascript community?
+    * What is the sequence? Why are pairs enough to implement sequences?
+    * Why does mapping increases the readability of the program?
+    * How to represent lists with pairs?
+    * How to represent arbitrary deep and arbitrary wide trees with pairs?
+    * What is the common way how to rewrite procedures for list to make them work with trees?
+    * What are the conventional interfaces? What is the relation between conventional interface and abstraction barrier?
+    * What are folds? What's difference between left and right folds?
+    * What property should combining function have to make results indistinguishable both for left and right folds?
+    * What is the flatmap? When it can be useful?
+    * What is the stratified design? What is the difference between stratified design and abstraction barriers?
+    * What role do primitives play in stratified design?
+    * Why does stratified design make the program robust?
+
+  * [2.3 Symbolic Data](http://mitpress.mit.edu/sites/default/files/sicp/full-text/book/book-Z-H-16.html#%_sec_2.3)
+
+    * What are the sets?
+    * What is the binary tree? When the binary tree is balanced?
+    * What is the difference in implementing sets as unordered lists, ordered lists, binary tree?
+    * What is a set of records and when it can be useful?
+    * What is the Huffman Encoding Trees?
+
+  * [2.4 Multiple Representations for Abstract Data](http://mitpress.mit.edu/sites/default/files/sicp/full-text/book/book-Z-H-17.html#%_sec_2.4)
+
+    * Why keeping multiple representations for abstract data in the same system can be useful?
+    * What is the additivity? How is this property related to the representations for abstract data?
+    * What are the generic procedures?
+    * When rectangular and when polar form of the complex numbers are useful?
+    * What is the type tag? When it can be useful?
+    * Explain when to strip off and when to attach tags to the abstract data?
+    * What is the dispatching on type? Why this approach is not additive?
+    * What is the data-directed programming? Why this approach requires the global mutable table?
+    * Why data-directed approach helps us to avoid procedure names conflict?
+    * How apply-generic from this chapter works?
+    * What is the message-passing approach?
+    * Compare "dispatching on type", "data-directed" and "message-passing" approaches.
+
+  * [2.5 Systems with Generic Operations](http://mitpress.mit.edu/sites/default/files/sicp/full-text/book/book-Z-H-18.html#%_sec_2.5)
+
+    * Explain the idea of Generic Operations?
+    * Why two-level tags are used in the example of the generic arithmetic operations package?
+    * Why implementing each procedure for each possible type combinations is cumbersome?
```

Comment:
Why is implementing each procedure for each possible type combinations cumbersome?

Reply:
**is** cumbersome